### PR TITLE
Earth - Olga

### PR DESCRIPTION
**Pull request by OlgaSe**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | Classes that handle the business logic of the app. A database.
Describe in your own words what the Controller is doing in Rails | Controller is controlling the flow of the app (brains)
Describe in your own words what the View is doing in Rails | Views are rendering the data and handle the logic of displaying it.
Describe an edge-case controller test you wrote | For complete action method, the edge case was responding with not found if task is not exist.
What is the purpose of using strong params? (i.e. the params method in the controller) | For security reason, to prevent passing the data from forms directly to the model instance. As well as, using strong params makes easier to read by shorting list of parameters received from user and verifies that all required parameters were received.
How are Rails migrations related to Rails models? |  Rails migration is a change of database schema, which is a structure of a database(model).
Describe one area of Rails that are still unclear on | there is no big picture about all processes that going in the app and confuse with syntax.

### CODE REVIEW

File: app/views/tasks/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+<h2>This is edit page</h2>
```
Comment (beccaelenzil) [inline app/views/tasks/edit.html.erb:1]: You've got almost exactly the same code here to render a form as you do in `new`. You should dry this up with a view partial.

File: app/views/tasks/edit.html.erb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+<div>
+    <%= f.label :description %>
+    <%= f.text_field :description %>
+</div>
+
+<div>
+    <%= f.label :completed_at %>
+    <%= f.text_field :completed_at %>
+</div>
```
Comment (beccaelenzil) [inline app/views/tasks/edit.html.erb:19]: Consider removing this from the form so that it is just set with the `mark_complete` method.

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    def complete
+        task_id = params[:id]
+        @task = Task.find_by(id: task_id)
+
+        if @task.nil?
+            head :not_found
+            return
+        elsif @task.update(
+            completed_at: Time.now.strftime("%m/%d/%Y")
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:100]: You should add functionality to mark a task incomplete.

Comment (beccaelenzil) [reply app/controllers/tasks_controller.rb:100]: Consider how making two separate methods, for instance `mark_complete` and `mark_incomplete` would make this action idempotent.

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+    def update
+        @task = Task.find_by(id: params[:id])
+        if @task.nil?
+            head :not_found
+            return
+        elsif @task.update(
+            name: params[:task][:name], 
+            description: params[:task][:description], 
+            completed_at: params[:task][:completed_at]
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:68]: consider using strong params

File: app/controllers/tasks_controller.rb [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+class TasksController < ApplicationController
+    def index
+        @tasks = Task.all
+    end
+
+    def show 
+        task_id = params[:id]
+        
+        begin 
+            @task = Task.find(task_id)
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:12]: Instead of using `find`, consider using `find_by`. `find_by` return nil if the object is not found.

File: app/controllers/tasks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    def edit
+        @task = Task.find_by(id: params[:id])
+
+        task_id = params[:id]
+        begin 
+            @task = Task.find(task_id)
+        rescue ActiveRecord::RecordNotFound
+            @task = nil
+        end
```
Comment (beccaelenzil) [inline app/controllers/tasks_controller.rb:52]: see comment for show action

**Review by beccaelenzil — COMMENTED** [review]
# Task List




## Major Learning Goals/Code Review





| Criteria                                           | yes/no, and optionally any details/lines of code to reference |
| -------------------------------------------------- | ------------------------------------------------------------- |
| At least 6 commits with meaningful commit messages | ✔️                                                            |
| Routes follow RESTful conventions                  | ✔️                                                            |
| Uses named routes (like `_path`)                   | ✔️                                                            |
| Creates Models and migrations                      | ✔️                                                            |
| Creates styled views                               | Not required                                                            |
| Handles errors like nonexistant tasks              | ✔️ Consider redirect rather than head :not_found                                                            |
| Uses `form_with` to render forms in Rails          | ✔️                                                            |

## Functional Requirements/Manual Testing

| Functional Requirement                                                         | yes/no |
| ------------------------------------------------------------------------------ | ------ |
| Successfully handles index & show                                              | ✔️     |
| index & show tests pass                                                        | ✔️     |
| Successfully handles: New, Create                                              | ✔️     |
| New, Create tests pass                                                         | ✔️     |
| Successfully handles: Edit, Update                                             | ✔️     |
| Edit, Update tests pass with valid & invalid task ids                          | ✔️     |
| Successfully handles: Destroy, Task Complete                                   | Add functionality for marking incomplete     |
| Tests for Destroy & Task Complete include tests for valid and invalid task ids | ✔️     |

## Overall Feedback

Good work overall on this project. You've created your first web app in rails, nice work! Your code is well written and well tested. You've met the learning goals of creating a rails app using rails conventions, including writing custom routes and methods. On future projects you will have the opportunity to work on styles and using features such as partial views and strong params. I've left a few in line comments for you to review. Keep up the hard work! 

| Overall Feedback                | Criteria                                                                                                                                                                                      | yes/no |
| ------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 6+ in Functional Requirements                                                                                                                                            |✔️
| Yellow (Approaches Standards)   | 3+ in Code Review && 5+ in Functional Requirements, or the instructor judges that this project needs special attention                                                                        |
| Red (Not at Standard)           | 0-2 in Code Review or 0-4 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality              | Yes? |
| -------------------- | ---- |
| Perfect Indentation  | ✅    |
| Descriptive/Readable | ✅    |
| Logical/Organized    | ✅    |
